bpf/lib/nat.h

```diff
@@ -558,24 +577,90 @@ snat_v4_process(struct __ctx_buff *ctx, enum nat_dir dir,
 		return NAT_PUNT_TO_STACK;
 	};
 
-	if (snat_v4_can_skip(target, &tuple, dir, from_endpoint, icmp_echoreply))
+	if (snat_v4_nat_can_skip(target, &tuple, from_endpoint, icmp_echoreply))
 		return NAT_PUNT_TO_STACK;
-	ret = snat_v4_handle_mapping(ctx, &tuple, &state, &tmp, dir, off, target);
+	ret = snat_v4_nat_handle_mapping(ctx, &tuple, &state, &tmp, off, target);
 	if (ret > 0)
 		return CTX_ACT_OK;
 	if (ret < 0)
 		return ret;
 
-	return dir == NAT_DIR_EGRESS ?
-	       snat_v4_rewrite_egress(ctx, &tuple, state, off, ipv4_has_l4_header(ip4)) :
-	       snat_v4_rewrite_ingress(ctx, &tuple, state, off);
+	return snat_v4_rewrite_egress(ctx, &tuple, state, off, ipv4_has_l4_header(ip4));
+}
+
+static __always_inline __maybe_unused int
+snat_v4_rev_nat(struct __ctx_buff *ctx, const struct ipv4_nat_target *target)
+{
+	struct icmphdr icmphdr __align_stack_8;
+	struct ipv4_nat_entry *state;
+	struct ipv4_ct_tuple tuple = {};
+	void *data, *data_end;
+	struct iphdr *ip4;
+	struct {
+		__be16 sport;
+		__be16 dport;
+	} l4hdr;
+	__u64 off;
+	int ret;
+
+	build_bug_on(sizeof(struct ipv4_nat_entry) > 64);
+
+	if (!revalidate_data(ctx, &data, &data_end, &ip4))
+		return DROP_INVALID;
+
+	tuple.nexthdr = ip4->protocol;
+	tuple.daddr = ip4->daddr;
+	tuple.saddr = ip4->saddr;
+	tuple.flags = NAT_DIR_INGRESS;
+	off = ((void *)ip4 - data) + ipv4_hdrlen(ip4);
```

Comment:
Could we move this to an inlined helper function to reduce the boilerplate?

Reply:
Hum.. can you be more specific? It seems that may be related to a larger refactoring as not only nat.h is using this pattern. I can open an other PR to make that cleans if you are agree.

Reply:
I was suggesting to add a helper function to fill the tuple, as the tuple extraction code in `snat_v4_nat()` and `snat_v4_rev_nat()` looks the same (same for the v6).

Reply:
Honestly even aligned I'm a bit mitigated doing this change, I understand that you want to avoid any redundant part but isn't that too much at that point ? :-) Please let me know, if you really want it I will make this to happen like a charm.

Reply:
I agree with Sahid here. It won't be trivial to share the code between IPv4 and IPv6 and we already have such code inlined in other places (e.g., bpf_lxc).

Reply:
I meant two helper functions - one for ipv4, one for ipv6. For instance, both `snat_v4_nat()` and `snat_v4_rev_nat()` share the same tuple extraction code. Any duplicate code is a potential source for bugs.